block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -173,28 +189,52 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: BlockStore: RaiseIO
                      Option(children).pure[F]
                    case None =>
                      for {
-                       blockOpt <- BlockStore[F].get(blockHash)
-                       result <- blockOpt match {
-                                  case Some(block) =>
-                                    val number = blockNumber(block)
-                                    if (number >= sortOffset) {
+                       blockNumberOpt <- getBlockNumber(blockHash)
+                       result <- blockNumberOpt match {
+                                  case Some(blockNumber) =>
+                                    if (blockNumber >= sortOffset) {
                                       none[Set[BlockHash]].pure[F]
                                     } else {
                                       lock.withPermit(
                                         for {
-                                          oldDagInfo <- loadCheckpoint(number)
+                                          oldDagInfo <- loadCheckpoint(blockNumber)
                                         } yield oldDagInfo.flatMap(_.childMap.get(blockHash))
                                       )
                                     }
-                                  case None => none[Set[BlockHash]].pure[F]
+                                  case None =>
+                                    none[Set[BlockHash]].pure[F]
                                 }
                      } yield result
                  }
       } yield result
     def lookup(blockHash: BlockHash): F[Option[BlockMetadata]] =
-      dataLookup.get(blockHash).pure[F]
+      dataLookup.get(blockHash) match {
+        case Some(blockMetadata) =>
+          Option(blockMetadata).pure[F]
+        case None =>
+          for {
+            blockNumberOpt <- getBlockNumber(blockHash)
+            result <- blockNumberOpt match {
+                       case Some(blockNumber) =>
+                         if (blockNumber >= sortOffset) {
+                           none[BlockMetadata].pure[F]
+                         } else {
+                           lock.withPermit(
+                             for {
+                               oldDagInfo <- loadCheckpoint(blockNumber)
+                             } yield oldDagInfo.flatMap(_.dataLookup.get(blockHash))
```

Comment:
I think it's possible to skip the `for`:
`loadCheckpoint(blockNumber).map(_.flatMap(_.dataLookup.get(blockHash)))`